src/Lecture3.hs

```diff
@@ -120,14 +139,19 @@ If you combine multiple rewards, the final reward will contain a
 special prize if at least one of the rewards is special.
 -}
 data Reward = Reward
-    { rewardGold    :: Gold
-    , rewardSpecial :: Bool
+    { rewardGold    :: !Gold
+    , rewardSpecial :: !Bool -- TODO replace Bool by Any
     } deriving (Show, Eq)
 
 instance Semigroup Reward where
-
+  (<>) :: Reward -> Reward -> Reward
+  (Reward lhsGold lhsSpecial) <> (Reward rhsGold rhsSpecial) = Reward gold special
+    where gold = lhsGold <> rhsGold
+          special = getAny $ Any lhsSpecial <> Any rhsSpecial
```

Comment:
You can use `||` in this task for simplicity to avoid `Any` wrapping and unwrapping. In the end, it comes to the final trade-off. Unwrapping `Any` might be inconvenient or it might not have the required instances so `Bool` works better.

But there're also advanced ways to derive `Semigroup` instances in Haskell, so if you specify `Any`, you could get the desired behaviour of `Semigroup` without the need to write it manually 🙂